### PR TITLE
feat: Satsuma integration for /catalog endpoint

### PR DESCRIPTION
**Pull request by juanmahidalgo**
Adds a new endpoint `/catalog` that accept most of the already available filters for Item and NFTs and returns the items for the new unified market 

TODO:
- Update README

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -48,6 +48,8 @@ import { IProcessedPricesComponent } from './ports/prices/types'
 import { IStatsComponent } from './ports/stats/types'
 import { IOwnerDataComponent } from './ports/owner/types'
 import { ItemOptions } from './ports/items/types'
+import { IPgComponent } from '@well-known-components/pg-component'
```
Comment (kevinszuchet) [inline src/types.ts:51]: Would you mind moving this up?

File: src/adapters/handlers/items.ts [hunk 2]
```diff
@@ -1,66 +1,22 @@
...
-    )
-    const emoteGenders = params.getList<GenderFilterOption>(
-      'emoteGender',
-      GenderFilterOption
-    )
-    const emotePlayMode = params.getList<EmotePlayMode>(
-      'emotePlayMode',
-      EmotePlayMode
-    )
+    const { contractAddresses, itemId, urns, ...restOfFilters } =
```
Comment (kevinszuchet) [inline src/adapters/handlers/items.ts:17]: Why is this necessary? For example, the `urns` are not used after deconstruction.

File: src/ports/catalog/component.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+} from './types'
+import {
+  getSubgraphNameForNetwork,
+  getCatalogQuery,
+  fromCollectionsItemDbResultToCatalogItem,
+} from './utils'
+import {
+  getCollectionsChainId,
+  getMarketplaceChainId,
+} from '../../logic/chainIds'
```
Comment (kevinszuchet) [inline src/ports/catalog/component.ts:17]: Would you mind moving this up?

File: src/ports/catalog/component.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import { getLatestSubgraphSchema } from './queries'
+
+export function createCatalogComponent(options: {
+  database: IPgComponent
```
Comment (kevinszuchet) [inline src/ports/catalog/component.ts:21]: WDYT about renaming the property `database` to `satsumaDatabase` or something similar to explicit that we are not using our DB?

File: src/ports/catalog/component.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        getCatalogQuery(
+          schemas.reduce((acc, curr) => ({ ...acc, ...curr }), {}),
+          filters
+        )
+      )
+      catalogItems = results.rows.map((res) =>
+        fromCollectionsItemDbResultToCatalogItem(res, network)
+      )
+    } catch (e) {
+      console.log(e)
```
Comment (kevinszuchet) [inline src/ports/catalog/component.ts:64]: Why not throw an error to show the message in the marketplace?

File: src/ports/catalog/queries.ts [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+} from './types'
+import {
+  EmoteCategory,
+  EmotePlayMode,
+  GenderFilterOption,
+  NFTCategory,
+  WearableCategory,
+  WearableGender,
+} from '@dcl/schemas'
+import { FragmentItemType } from '../items/types'
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:16]: Would you mind re-sorting the imports?

```ts
import SQL, { SQLStatement } from 'sql-template-strings'
import {
  EmoteCategory,
  EmotePlayMode,
  GenderFilterOption,
  NFTCategory,
  WearableCategory,
  WearableGender,
} from '@dcl/schemas'
import { FragmentItemType } from '../items/types'
import {
  CatalogFilters,
  CatalogQueryFilters,
  CatalogSortBy,
  CatalogSortDirection,
} from './types'
```

File: src/ports/catalog/queries.ts [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+export function getOrderBy(filters: CatalogFilters) {
+  const { sortBy, sortDirection, onlyListing, isOnSale } = filters
+  const sortByParam = sortBy ?? CatalogSortBy.NEWEST
+  const sortDirectionParam = sortDirection ?? CatalogSortDirection.ASC
+
+  // When seeing "Not for sale", the only sort available is the Newest one
+  if (isOnSale === false) {
+    if (sortByParam === CatalogSortBy.NEWEST) {
+      return `ORDER BY items.first_listed_at ${sortDirectionParam}\n`
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:36]: I think this is already contemplated a few lines below in the `case CatalogSortBy.NEWEST:` (line 46).

We may just return an empty string if the `isOnSale` filter is `false` and the `sortByParam` is not `NEWEST`. WDYT?

```ts
if (isOnSale === false && sortByParam !== CatalogSortBy.NEWEST) return ''
```

File: src/ports/catalog/queries.ts [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+    }
+    return ''
+  }
+
+  let sortByQuery:
+    | SQLStatement
+    | string = `ORDER BY items.first_listed_at ${sortDirectionParam}\n`
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:43]: Would you mind using the `default` case instead of initializing the `sortByQuery` variable? Just to be more explicit.

File: src/ports/catalog/queries.ts [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+  return sortByQuery
+}
+
+export const addQuerySortAndPagination = (
+  query: SQLStatement,
+  filters: CatalogQueryFilters
+) => {
+  const { sortBy, sortDirection, limit, offset } = filters
+  if (sortBy && sortDirection) {
+    query.append(getOrderBy(filters)) // VER TODO
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:73]: Is this comment still relevant?

File: src/ports/catalog/queries.ts [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+        .append(wearableCategory)
+        .append(SQL`'`)
+    : SQL``
+}
+
+export const getEmoteCategoryWhere = (emoteCategory: EmoteCategory) => {
+  return EmoteCategory.validate(emoteCategory)
+    ? SQL`metadata_emote.category = '`.append(emoteCategory).append(SQL`'`)
+    : SQL``
+}
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:116]: I think `getWearableCategoryWhere` and `getEmoteCategoryWhere` have similar behavior. We could write a new function called `getCategoryWhere` or `getMetadataCategoryWhere` to generalize the duplicated logic. WDYT?

File: src/ports/catalog/queries.ts [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+export const getOnlyListingsWhere = () => {
+  return SQL`(items.search_is_store_minter = false OR (items.search_is_store_minter = true AND available = 0)) AND listings_count > 1`
+}
+
+export const getOnlyMintingWhere = () => {
+  return SQL`items.search_is_store_minter = true AND available > 0`
+}
+
+export const getCollectionsQueryWheres = (filters: CatalogFilters) => {
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:191]: Is `getCollectionsQueryWheres` a typo?

```suggestion
export const getCollectionsQueryWhere = (filters: CatalogFilters) => {
```

File: src/ports/catalog/queries.ts [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+    .append(
+      `.item_active AS items 
+            LEFT JOIN (
+              SELECT 
+                nft.item, 
+                COUNT(orders.id) AS listings_count,
+                MIN(orders.price) AS min_price,
+                MAX(orders.price) AS max_price,
+                COUNT(DISTINCT nft.owner) as owners_count
+                MAX(orders.updated_at) AS max_order_created_at
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:293]: This statement uses the column `updated_at` and renames it as `max_order_created_at `. Is this on purpose, or is it an error?

File: src/ports/catalog/queries.ts [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+      `.nft_active AS nft 
+                JOIN 
+              `
+    )
+    .append(schemaVersion)
+    .append(
+      `.order_active AS orders ON orders.nft = nft.id 
+              WHERE 
+                orders.status = 'open' 
+                AND orders.expires_at < 253378408747000
```
Comment (kevinszuchet) [inline src/ports/catalog/queries.ts:307]: We could store the magic timestamp in a constant with a human-friendly name, so our future selves would be proud of us. WDYT?

File: src/ports/catalog/types.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
+// import { ListingStatus } from '@dcl/schemas'
```
Comment (kevinszuchet) [inline src/ports/catalog/types.ts:1]: Can we delete this comment?

File: src/tests/ports/catalog-queries.spec.ts [hunk 15]
```diff
@@ -0,0 +1,399 @@
+import SQL, { SQLStatement } from 'sql-template-strings'
+import { test } from '../components'
```
Comment (kevinszuchet) [inline src/tests/ports/catalog-queries.spec.ts:2]: Would you mind moving this down?

File: src/adapters/handlers/items.ts (hunk 2 above)
Comment (juanmahidalgo) [reply src/adapters/handlers/items.ts:17]: They are used here: https://github.com/decentraland/nft-server/blob/master/src/adapters/handlers/items.ts#L70. Basically I'm leaving those 3 out of the rest so they can be used in that line

File: src/ports/catalog/component.ts (hunk 4 above)
Comment (juanmahidalgo) [reply src/ports/catalog/component.ts:21]: I think this catalog component can be re-used in the future by another database (let's say our own one) so at this point, it makes sense it's agnostic to who owns it. At the component level, when the database component is created, I called it `satsumaDatabase` https://github.com/decentraland/nft-server/pull/266/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R686. Does it make sense?

File: src/ports/catalog/queries.ts (hunk 10 above)
Comment (juanmahidalgo) [reply src/ports/catalog/queries.ts:116]: I think the abstraction will be more complex than having 1 line duplicated. Specially the piece of validating the enum, if it recieves one enum or the other, we can't validate which is the one being called, it would requiere something like another parameter with a hardcoded string and having some ternaries for the validation and the SQL piece. I tried but it think that sometimes is better to have more code than having the wrong abstraction

File: src/ports/catalog/queries.ts (hunk 11 above)
Comment (juanmahidalgo) [reply src/ports/catalog/queries.ts:191]: I was intentional :p but `where` sounds a little bit better

File: src/ports/catalog/component.ts [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    client.release()
+    return catalogItems
```
Comment (kevinszuchet) [inline src/ports/catalog/component.ts:71]: Should this be enhanced with the favorites data like in the items source?

File: src/ports/catalog/component.ts [hunk 17]
```diff
@@ -20,10 +21,12 @@ import { getLatestSubgraphSchema } from './queries'
...
 export function createCatalogComponent(options: {
   database: IPgComponent
+  favoritesComponent: IFavoritesComponent
+  isFavoritesEnabled: boolean
 }): ICatalogComponent {
-  const { database } = options
+  const { database, favoritesComponent, isFavoritesEnabled } = options
 
-  async function fetch(filters: CatalogFilters) {
+  async function fetch(filters: CatalogOptions) {
```
Comment (kevinszuchet) [inline src/ports/catalog/component.ts:29]: In the other ports that also enhance the data with the picks stats, we deconstructed the filters so we have `{ pickedBy, ...filters}` and use the "real" filters to query what we need.

File: src/tests/ports/catalog.spec.ts [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+        }
+
+        dbClientQueryMock.mockResolvedValueOnce(latestSubgraphSchemaResponse)
+        dbClientQueryMock.mockResolvedValueOnce({
+          rows: [dbItemResponse],
+          rowCount: 1,
+        })
+      })
+
+      it('should fetch the data from one subgraph return the catalog items', async () => {
```
Comment (kevinszuchet) [inline src/tests/ports/catalog.spec.ts:125]:
```suggestion
      it('should fetch the data from one subgraph and return the catalog items', async () => {
```

**Review by kevinszuchet — APPROVED** [review]

File: src/tests/ports/catalog.spec.ts [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+        latestSubgraphSchemaResponse = {
+          rows: [
+            {
+              entity_schema: latestSchema,
+            },
+          ],
+          rowCount: 0,
+        }
+
+        dbItemResponse = {
```
Comment (kevinszuchet) [inline src/tests/ports/catalog.spec.ts:82]: Something from this object is breaking the tests. Would you TAL please?

**Review by kevinszuchet — APPROVED** [review]

**Review by kevinszuchet — APPROVED** [review]
